dvc/remote/index.py

```diff
@@ -0,0 +1,78 @@
+import logging
+import pathlib
+import pickle
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteIndex(object):
+    """Class for locally indexing remote checksums.
+
+    Args:
+        repo: repo for this remote
+    """
+
+    INDEX_SUFFIX = ".idx"
+
+    def __init__(self, repo, name):
+        if repo and hasattr(repo, "index_dir") and name:
+            self.path = pathlib.Path(repo.index_dir).joinpath(
+                "{}{}".format(name, self.INDEX_SUFFIX)
+            )
+        else:
+            self.path = None
+        self._checksums = set()
+        self.load()
+
+    def __iter__(self):
+        return iter(self._checksums)
+
+    @property
+    def checksums(self):
+        return self._checksums
+
+    def load(self):
+        """(Re)load this index from disk."""
+        if self.path and self.path.exists():
+            try:
+                with open(self.path, "rb") as fd:
+                    self._checksums = pickle.load(fd)
+            except IOError:
+                logger.error(
+                    "Failed to load remote index from '{}'".format(self.path)
+                )
+
+    def save(self):
+        """Save this index to disk."""
+        if self.path:
+            try:
+                with open(self.path, "wb") as fd:
+                    pickle.dump(self._checksums, fd)
```

Comment:
What are the compatibility guarantees for pickle, btw? Like with python versions and backward/forward compatibility.

Reply:
By default, any current python (up through 3.8) writes a protocol version that is compatible with >=3.4. There is a new protocol version that's only readable in 3.8, but it's not the default.

I'm also just using pickle as a WIP placeholder right now, I don't think it's actually the right solution here